src/main/java/htsjdk/samtools/SamFiles.java

```diff
@@ -19,61 +21,73 @@
-                System.err.println("The index file " + canonicalIndexFile.getPath()
+            final Path canonicalSamPath = samPath.toRealPath();
+            final Path canonicalIndexPath = lookForIndex(canonicalSamPath);
+            if ( canonicalIndexPath != null) {
+                System.err.println("The index file " + canonicalIndexPath.toAbsolutePath()
                         + " was found by resolving the canonical path of a symlink: "
-                        + samFile.getPath() + " -> " + samFile.getCanonicalPath());
+                        + samPath.toAbsolutePath() + " -> " + samPath.toRealPath());
             }
-            return canonicalIndexFile;
+            return canonicalIndexPath;
         } catch (IOException e) {
             return null;
         }
     }
 
-    private static File lookForIndex(final File samFile) {// If input is foo.bam, look for foo.bai
-        File indexFile;
-        final String fileName = samFile.getName();
+    private static Path lookForIndex(final Path samPath) {// If input is foo.bam, look for foo.bai
+        Path indexPath;
+        final String fileName = samPath.getFileName().toString();
         if (fileName.endsWith(BamFileIoUtils.BAM_FILE_EXTENSION)) {
             final String bai = fileName.substring(0, fileName.length() - BamFileIoUtils.BAM_FILE_EXTENSION.length()) + BAMIndex.BAMIndexSuffix;
-            indexFile = new File(samFile.getParent(), bai);
-            if (indexFile.isFile()) {
-                return indexFile;
+            indexPath = samPath.resolveSibling(bai);
+            if (Files.isRegularFile(indexPath)) {
+                return indexPath;
             }
 
 
         } else if (fileName.endsWith(CramIO.CRAM_FILE_EXTENSION)) {
             final String crai = fileName.substring(0, fileName.length() - CramIO.CRAM_FILE_EXTENSION.length()) + CRAIIndex.CRAI_INDEX_SUFFIX;
-            indexFile = new File(samFile.getParent(), crai);
-            if (indexFile.isFile()) {
-                return indexFile;
+            indexPath = samPath.resolveSibling(crai);
+            if (Files.isRegularFile(indexPath)) {
+                return indexPath;
             }
 
-            indexFile = new File(samFile.getParent(), samFile.getName() + CRAIIndex.CRAI_INDEX_SUFFIX);
-            if (indexFile.isFile()) {
-                return indexFile;
+            indexPath = samPath.resolveSibling(fileName + CRAIIndex.CRAI_INDEX_SUFFIX);
+            if (Files.isRegularFile(indexPath)) {
+                return indexPath;
             }
         }
 
         // If foo.bai doesn't exist look for foo.bam.bai
-        indexFile = new File(samFile.getParent(), samFile.getName() + BAMIndex.BAMIndexSuffix);
-        if (indexFile.isFile()) {
-            return indexFile;
+        indexPath = samPath.resolveSibling(fileName + BAMIndex.BAMIndexSuffix);
+        if (Files.isRegularFile(indexPath)) {
+            return indexPath;
         }
```

Comment:
Add comment that this code is here and not in the `if (fileName.endsWith(BamFileIoUtils.BAM_FILE_EXTENSION)) {` block because we want to fall back to looking for bai's on cram files as well (at least, that is my interpretation of why this is here instead of above)